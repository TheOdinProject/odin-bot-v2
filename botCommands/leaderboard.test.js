const axios = require('axios');
const commands = require('./leaderboard');
const { generateLeaderData } = require('./mockData');
/* eslint-disable */
/* eslint max-classes-per-file: ["error", 2] */

class GuildMembersMock {
  members;

  cache;

  constructor(users) {
    this.members = users;
    this.cache = {
      get: (id) => users.filter((member) => member.discord_id === id)[0],
    };
  }
}

class GuildMock {
  members;

  member;

  constructor(users) {
    this.members = new GuildMembersMock(users);
    this.member = (user) => users.filter((member) => member === user)[0];
  }
}

describe('!leaderboard', () => {
  describe('regex', () => {
    it.each([
      ['!leaderboard'],
      ['<@!123456789> !leaderboard'],
      ['!leaderboard n=10 start=30'],
      ['!leaderboard n=20 start=50'],
      ['!leaderboard n=10'],
      ['!leaderboard start=30'],
    ])('correct strings trigger the callback', (string) => {
      expect(commands.leaderboard.regex.test(string)).toBeTruthy();
    });

    it.each([
      ['!leaderboad'],
      [''],
      [' '],
      [' !'],
      ['!lead'],
      ['leaderboard'],
      ['!le'],
      ['!leaderboards'],
      ['```function("!leaderboard", () => {}```'],
      ['!leader'],
      ['<@!123456789> ! leaderboard'],
      ['<@!123456789> !leaderbard'],
      ['!leaderbord n=10 start=30'],
    ])("'%s' does not trigger the callback", (string) => {
      expect(commands.leaderboard.regex.test(string)).toBeFalsy();
    });

    it.each([
      ['Check this out! !leaderboard'],
      ["Don't worry about !leaderboard"],
      ['Hey <@!123456789>, !leaderboard'],
      ['!<@!123456789> ^ !me !leaderboard !tests$*'],
    ])("'%s' - command can be anywhere in the string", (string) => {
      expect(commands.leaderboard.regex.test(string)).toBeTruthy();
    });

    it.each([
      ['@user!leaderboard'],
      ["it's about!leaderboard"],
      ['!leaderboardisanillusion'],
      ['!leaderboard!'],
      ['!leaderboard*'],
      ['!leaderboard...'],
    ])(
      "'%s' - command should be its own word!group - no leading or trailing characters",
      (string) => {
        expect(commands.leaderboard.regex.test(string)).toBeFalsy();
      }
    );
  });

  describe('callback', () => {
    const members = generateLeaderData(5);

    axios.get = jest.fn();
    axios.get.mockResolvedValue({
      data: members,
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(0, 5),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=5 start=1',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(0, 3),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=3 start=1',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(2, 4),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=2 start=3',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(2),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard start=3',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(0, 2),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=2',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(0, 2),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=2 start=wtf',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(2),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=wtf start=3',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(9998, 25),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=25 start=9999',
        })
      ).toMatchSnapshot();
    });
  });
});
describe('!leaderboard', () => {
  describe('regex', () => {
    it.each([
      ['!leaderboard'],
      ['<@!123456789> !leaderboard'],
      ['!leaderboard n=10 start=30'],
      ['!leaderboard n=20 start=50'],
      ['!leaderboard n=10'],
      ['!leaderboard start=30'],
    ])('correct strings trigger the callback', (string) => {
      expect(commands.leaderboard.regex.test(string)).toBeTruthy();
    });

    it.each([
      ['!leaderboad'],
      [''],
      [' '],
      [' !'],
      ['!lead'],
      ['leaderboard'],
      ['!le'],
      ['!leaderboards'],
      ['```function("!leaderboard", () => {}```'],
      ['!leader'],
      ['<@!123456789> ! leaderboard'],
      ['<@!123456789> !leaderbard'],
      ['!leaderbord n=10 start=30'],
    ])("'%s' does not trigger the callback", (string) => {
      expect(commands.leaderboard.regex.test(string)).toBeFalsy();
    });

    it.each([
      ['Check this out! !leaderboard'],
      ["Don't worry about !leaderboard"],
      ['Hey <@!123456789>, !leaderboard'],
      ['!<@!123456789> ^ !me !leaderboard !tests$*'],
    ])("'%s' - command can be anywhere in the string", (string) => {
      expect(commands.leaderboard.regex.test(string)).toBeTruthy();
    });

    it.each([
      ['@user!leaderboard'],
      ["it's about!leaderboard"],
      ['!leaderboardisanillusion'],
      ['!leaderboard!'],
      ['!leaderboard*'],
      ['!leaderboard...'],
    ])(
      "'%s' - command should be its own word/group - no leading or trailing characters",
      (string) => {
        expect(commands.leaderboard.regex.test(string)).toBeFalsy();
      }
    );
  });

  describe('callback', () => {
    const members = generateLeaderData(5);

    axios.get = jest.fn();
    axios.get.mockResolvedValue({
      data: members,
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(0, 5),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=5 start=1',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(0, 3),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=3 start=1',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(2, 4),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=2 start=3',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(2),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard start=3',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(0, 2),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=2',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(0, 2),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=2 start=wtf',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(2),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=wtf start=3',
        })
      ).toMatchSnapshot();
    });

    it('returns correct output', async () => {
      axios.get.mockResolvedValue({
        data: members.slice(9998, 25),
      });

      expect(
        await commands.leaderboard.cb({
          guild: new GuildMock(members),
          content: '!leaderboard n=25 start=9999',
        })
      ).toMatchSnapshot();
    });
  });
});
