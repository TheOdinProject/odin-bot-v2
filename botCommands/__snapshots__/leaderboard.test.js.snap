// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`!leaderboard callback returns correct output 1`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 2`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 3`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "3 - user102 [998 points] 
4 - user103 [997 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 4`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 5`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 6`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "Be the first to earn a point!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 7`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "Be the first to earn a point!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 8`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "Be the first to earn a point!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 9`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 10`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 11`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "3 - user102 [998 points] 
4 - user103 [997 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 12`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 13`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 14`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "Be the first to earn a point!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 15`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "Be the first to earn a point!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`!leaderboard callback returns correct output 16`] = `
Object {
  "embeds": Array [
    Object {
      "color": 13407555,
      "fields": Array [
        Object {
          "name": "In Odin We Trust",
          "value": "Be the first to earn a point!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;
